src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+{
+    std::size_t first = peer.find_first_of(":");
+    std::size_t last = peer.find_last_of(":");
+    std::string peer_clean(peer);
+    if(first != last)
+    {
+        peer_clean = peer.substr(first+1);
+    }
+    return peer_clean;
+}
+
+} //namespace
+
+class GRPCServerImpl final {
+
+    private:
+
+    //list of current RPC requests being processed or listened for
+    std::list<std::shared_ptr<Processor>> requests_;
+
+    //CompletionQueue returns events that have occurred, or events that have
+    //been cancelled
+    std::unique_ptr<grpc::ServerCompletionQueue> cq_;
+
+    //The gRPC service defined by the .proto files
+    rpc::v1::XRPLedgerAPIService::AsyncService service_;
+
+    //gRPC server
+    std::unique_ptr<grpc::Server> server_;
+
+    //referernce to ripple::Application
+    Application& app_;
+
+    //address of where to run the server
+    //this is the default value, if ip and port are not in config
+    std::string server_address_ = "0.0.0.0:50051";
+
+
+    //typedef for function to bind a listener
+    template <class Request, class Response>
+    using BindListener = std::function<void(
+            rpc::v1::XRPLedgerAPIService::AsyncService&,
+            grpc::ServerContext*,
+            Request*,
+            grpc::ServerAsyncResponseWriter<Response>*,
+            grpc::CompletionQueue*,
+            grpc::ServerCompletionQueue*,
+            void*)>;
+
+    //typedef for actual handler (that populates a response)
+    template <class Request, class Response>
+    using Handler = std::function<std::pair<Response,grpc::Status>(
+            RPC::ContextGeneric<Request>&)>;
+
+    public:
+    ~GRPCServerImpl() {}
+
+    GRPCServerImpl(Application& app);
+  
+    void shutdown()
+    {
+        server_->Shutdown();
+        // Always shutdown the completion queue after the server.
+        cq_->Shutdown(); 
+    }
+  
+    //setup the server and listeners
+    void start();
+
+    //the main event loop
+    void handleRpcs();
+  
+    private:
+  
+    // Class encompasing the state and logic needed to serve a request.
+    template <class Request,class Response>
+    class CallData :
+        public Processor, 
+        public std::enable_shared_from_this<CallData<Request,Response>>
+    {
+        private:
+        // The means of communication with the gRPC runtime for an asynchronous
+        // server.
+        rpc::v1::XRPLedgerAPIService::AsyncService& service_;
+
+        // The producer-consumer queue for asynchronous server notifications.
+        grpc::ServerCompletionQueue& cq_;
+
+        // Context for the rpc, allowing to tweak aspects of it such as the use
+        // of compression, authentication, as well as to send metadata back to
+        // the client.
+        grpc::ServerContext ctx_;
+
+        // Possible states of the RPC
+        enum CallStatus { PROCESSING, FINISH };
+        // The current serving state.
+        CallStatus status_;
```

Comment:
I think a `bool finished_` or somesuch is clearer. I understand this started with more states, and we pared it down to two, but now that it's two states I'd just use a bool. Tho I guess I don't object too strongly if other reviews are OK with `enum`.

Reply:
I think a bool would be clearer as well, will change

Reply:
changed to bool